app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt

```diff
@@ -4,175 +4,123 @@
         }
     }
 
-    override fun openInNewTab(item: BookmarkNode) {
+    override fun onOpenInNormalTab(item: BookmarkNode) {
         require(item.type == BookmarkNodeType.ITEM)
-        item.url?.let { url ->
-            activity?.browsingModeManager?.mode =
-                BrowsingModeManager.Mode.Normal
-            activity?.openToBrowserAndLoad(
-                searchTermOrURL = url,
-                newTab = true,
-                from = BrowserDirection.FromBookmarks
-            )
+        item.url?.let {
+            bookmarksController.handleOpeningBookmark(item, BrowsingModeManager.Mode.Normal)
             metrics.track(Event.OpenedBookmarkInNewTab)
         }
     }
 
-    override fun openInPrivateTab(item: BookmarkNode) {
+    override fun onOpenInPrivateTab(item: BookmarkNode) {
         require(item.type == BookmarkNodeType.ITEM)
-        item.url?.let { url ->
-            activity?.browsingModeManager?.mode =
-                BrowsingModeManager.Mode.Private
-            activity?.openToBrowserAndLoad(
-                searchTermOrURL = url,
-                newTab = true,
-                from = BrowserDirection.FromBookmarks
-            )
+        item.url?.let {
+            bookmarksController.handleOpeningBookmark(item, BrowsingModeManager.Mode.Private)
             metrics.track(Event.OpenedBookmarkInPrivateTab)
         }
     }
 
-    override fun delete(nodes: Set<BookmarkNode>) {
+    override fun onDelete(nodes: Set<BookmarkNode>) {
         val eventType = when (nodes.singleOrNull()?.type) {
             BookmarkNodeType.ITEM -> Event.RemoveBookmark
             BookmarkNodeType.FOLDER -> Event.RemoveBookmarkFolder
             BookmarkNodeType.SEPARATOR -> throw IllegalStateException("Cannot delete separators")
             null -> Event.RemoveBookmarks
         }
-        deleteBookmarkNodes(nodes, eventType)
+        bookmarksController.handleBookmarkDeletion(nodes, eventType)
     }
 
-    override fun backPressed() {
-        navController.popBackStack()
+    override fun onBackPressed() {
+        bookmarksController.handleBackPressed()
     }
 
-    override fun clickedSignIn() {
-        context.components.services.launchPairingSignIn(context, navController)
+    override fun onSignInPressed() {
+        bookmarksController.handleSigningIn()
+    }
+
+    override fun onSignedIn() {
+        viewModel.signedIn.postValue(true)
+    }
+
+    override fun onSignedOut() {
+        viewModel.signedIn.postValue(false)
+    }
+
+    override fun open(item: BookmarkNode) {
+        when (item.type) {
+            BookmarkNodeType.ITEM -> {
+                bookmarksController.handleBookmarkTapped(item)
+                metrics.track(Event.OpenedBookmark)
+            }
+            BookmarkNodeType.FOLDER -> bookmarksController.handleBookmarkExpand(item)
+            BookmarkNodeType.SEPARATOR -> throw IllegalStateException("Cannot open separators")
+        }.exhaustive
```

Comment:
nit: do we need this? It wasn't in the original `open` function.

Reply:
The `exhaustive` property?
Not necessarily needed but I feel it's a great sanity check for the future, to ensure all cases are properly treated.
https://proandroiddev.com/til-when-is-when-exhaustive-31d69f630a8b
That import though doesn't seem quite right...
If you agree to keep it maybe we could add it as a new extension propery in the `ext` package?

Reply:
Yeah, this seems like a great addition to the ext package! Thanks for clearing this up :)

Reply:
Found there is a [Do exhaustive](https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/lib/Do.kt) method available and already used in a few places for the same end result. Will go with that and avoid creating a new `T.exhaustive` method.